hphp/runtime/vm/jit/tc-relocate.cpp

```diff
@@ -518,18 +531,30 @@ bool relocateNewTranslation(TransLoc& loc, CodeCache::View cache,
   }
 
   if (asm_count) {
-    x64::adjustForRelocation(rel);
-    x64::adjustMetaDataForRelocation(rel, nullptr, fixups);
-    x64::adjustCodeForRelocation(rel, fixups);
+    adjustForRelocation(rel);
+    adjustMetaDataForRelocation(rel, nullptr, fixups);
+    adjustCodeForRelocation(rel, fixups);
   }
 
   if (debug) {
     auto clearRange = [](TCA start, TCA end) {
       CodeBlock cb;
       cb.init(start, end - start, "Dead code");
-      X64Assembler a {cb};
-      while (cb.available() >= 2) a.ud2();
-      if (cb.available() > 0) a.int3();
+      switch (arch()) {
+        case Arch::PPC64: {
+          ppc64_asm::Assembler a{cb};
+          a.emitTrap();
+          break;
+        }
+        case Arch::X64: {
+          X64Assembler a{cb};
+          a.emitTrap();
+          break;
+        }
+        case Arch::ARM:
+          not_implemented();
+          break;
+      }
```

Comment:
well, emitting only one ud2 didn't work. On X64 I see 74 tests failing with `--recycle-tc 5`. On `hphp/test/quick/yermo.php` I noticed this:
```
#6  0x0000000005e733ca in HPHP::jit::tc::(anonymous namespace)::clearTCMaps (start=0xb40055c "\217\r",
    end=0xb4005fa "\200}\370p\017\205\361\035\240\004H\213]\360\203{\f\201~\v\201{\f\377\377\377?v\002\017\v\200{\b\021\017\205\322\035\240\004H\270\350\002\300\337\377\177")
    at /root/hhvm/hphp/runtime/vm/jit/tc-recycle.cpp:145
145         x64::DecodedInstruction di(start);
(gdb x86) x /i start
   0xb40055c:   (bad)
```

I changed as you instructed:
```diff
diff --git c/hphp/runtime/vm/jit/tc-recycle.cpp w/hphp/runtime/vm/jit/tc-recycle.cpp
index 16e34dc..4810983 100644
--- c/hphp/runtime/vm/jit/tc-recycle.cpp
+++ w/hphp/runtime/vm/jit/tc-recycle.cpp
@@ -28,6 +28,7 @@
 #include "hphp/runtime/vm/jit/service-requests.h"
 #include "hphp/runtime/vm/jit/smashable-instr.h"
 #include "hphp/runtime/vm/jit/srcdb.h"
+#include "hphp/runtime/vm/jit/vasm-gen.h"
 
 #include "hphp/util/arch.h"
 #include "hphp/util/asm-x64.h"
@@ -210,22 +211,10 @@ void reclaimTranslation(TransLoc loc) {
     // Ensure no one calls into the function
     ITRACE(1, "Overwriting function\n");
     auto clearBlock = [] (CodeBlock& cb) {
-      switch (arch()) {
-        case Arch::PPC64: {
-          ppc64_asm::Assembler a{cb};
-          a.emitTrap();
-          break;
-        }
-        case Arch::X64: {
-          X64Assembler a{cb};
-          a.emitTrap();
-          break;
-        }
-        case Arch::ARM:
-          not_implemented();
-          break;
-      }
-      always_assert(!cb.available());
+      DataBlock db;
+      vwrap(cb, db, [] (Vout& v) {
+        v << ud2{};
+      });
     };
 
     CodeBlock main, cold, frozen;
diff --git c/hphp/runtime/vm/jit/tc-relocate.cpp w/hphp/runtime/vm/jit/tc-relocate.cpp
index d344bf8..195387b 100644
--- c/hphp/runtime/vm/jit/tc-relocate.cpp
+++ w/hphp/runtime/vm/jit/tc-relocate.cpp
@@ -30,6 +30,7 @@
 #include "hphp/runtime/vm/jit/print.h"
 #include "hphp/runtime/vm/jit/service-requests.h"
 #include "hphp/runtime/vm/jit/stub-alloc.h"
+#include "hphp/runtime/vm/jit/vasm-gen.h"
 
 #include "hphp/tools/hfsort/jitsort.h"
 
@@ -90,21 +91,10 @@ struct CodeSmasher {
     for (auto& e : entries) {
       CodeBlock cb;
       cb.init(e.first, e.second - e.first, "relocated");
-      switch (arch()) {
-        case Arch::PPC64: {
-          ppc64_asm::Assembler a{cb};
-          a.emitTrap();
-          break;
-        }
-        case Arch::X64: {
-          X64Assembler a{cb};
-          a.emitTrap();
-          break;
-        }
-        case Arch::ARM:
-          not_implemented();
-          break;
-      }
+      DataBlock db;
+      vwrap(cb, db, [] (Vout& v) {
+        v << ud2{};
+      });
     }
     okToRelocate = true;
   }
@@ -540,22 +530,10 @@ bool relocateNewTranslation(TransLoc& loc, CodeCache::View cache,
     auto clearRange = [](TCA start, TCA end) {
       CodeBlock cb;
       cb.init(start, end - start, "Dead code");
-      switch (arch()) {
-        case Arch::PPC64: {
-          ppc64_asm::Assembler a{cb};
-          a.emitTrap();
-          break;
-        }
-        case Arch::X64: {
-          X64Assembler a{cb};
-          a.emitTrap();
-          break;
-        }
-        case Arch::ARM:
-          not_implemented();
-          break;
-      }
-      always_assert(!cb.available());
+      DataBlock db;
+      vwrap(cb, db, [] (Vout& v) {
+        v << ud2{};
+      });
     };
 
     if (mainStartRel != loc.mainStart()) {
```